core/unit_test/TestArray.cpp

```diff
@@ -49,4 +49,49 @@ KOKKOS_FUNCTION constexpr bool test_array_structured_binding_support() {
 
 static_assert(test_array_structured_binding_support());
 
+template<typename L, typename R>
+constexpr bool is_equal(L const& l, R const& r) {
+    if (std::size(l) != std::size(r)) return false;
+
+    for (size_t i = 0; i != std::size(l); ++i) {
+        if (l[i] != r[i]) return false;
+    }
+
+    return true;
+}
+
+KOKKOS_FUNCTION constexpr bool test_array_ctad() {
+  constexpr int x = 10;
+  constexpr Kokkos::Array a{1, 2, 3, 5, x};
+  constexpr Kokkos::Array<int, 5> b{1, 2, 3, 5, x};
+
+  return std::is_same_v<decltype(a), decltype(b)> && is_equal(a, b);
+}
+
+static_assert(test_array_ctad());
+
+KOKKOS_FUNCTION constexpr bool test_to_array() {
+
+    // copies a string literal
+    auto a1 = Kokkos::to_Array("foo");
+    static_assert(a1.size() == 4);
+
+    // deduces both element type and length
+    auto a2 = Kokkos::to_Array({0, 2, 1, 3});
+    static_assert(std::is_same_v<decltype(a2), Kokkos::Array<int, 4>>);
+
+    // deduces length with element type specified
+    // implicit conversion happens
+    auto a3 = Kokkos::to_Array<long>({0, 1, 3});
+    static_assert(std::is_same_v<decltype(a3), Kokkos::Array<long, 3>>);
+
+    auto a4 = Kokkos::to_Array<std::pair<int, float>>(
+        {{3, 0.0f}, {4, 0.1f}, {4, 0.1e23f}});
+    static_assert(a4.size() == 3);
+
+    return true;
+}
```

Comment:
are we sure these guys get instantiated on device?

Reply:
Nothing in `TestArray.cpp` currently gets tested outside of compile time as it is in the `COMPILE_ONLY_SOURCES` section.  There are some minor tests for `Array` in `TestAggregate.hpp`.  Should I move those into a `TestArrayOps.hpp` and expand on them to cover `to_Array`?  Adding more unit tests for `Array` is also doable, but I'd rather do that as part of a separate PR.

Reply:
Added some runtime tests.

(There is still a possible ICPC issue, so this is still a draft)